Tests/MongoSwiftTests/MongoError+Equatable.swift

```diff
@@ -0,0 +1,87 @@
+             (.authenticationError(message: _), .authenticationError(message: _)):
+            return true
+        case let (.connectionError(message: _, errorLabels: lhsLabels),
+                  .connectionError(message: _, errorLabels: rhsLabels)):
+            return sortAndCompareOptionalArrays(lhs: lhsLabels, rhs: rhsLabels, cmp: { $0 < $1 })
+        default:
+            return false
+        }
+    }
+}
+
+extension ServerError: Equatable {
+    public static func == (lhs: ServerError, rhs: ServerError) -> Bool {
+        switch (lhs, rhs) {
+        case let (.commandError(code: lhsCode, message: _, errorLabels: lhsErrorLabels),
+                  .commandError(code: rhsCode, message: _, errorLabels: rhsErrorLabels)):
+            return lhsCode == rhsCode
+                    && sortAndCompareOptionalArrays(lhs: lhsErrorLabels, rhs: rhsErrorLabels, cmp: { $0 < $1 })
+        case let (.writeError(writeError: lhsWriteError, writeConcernError: lhsWCError, errorLabels: lhsErrorLabels),
+                  .writeError(writeError: rhsWriteError, writeConcernError: rhsWCError, errorLabels: rhsErrorLabels)):
+            return lhsWriteError == rhsWriteError
+                    && lhsWCError == rhsWCError
+                    && sortAndCompareOptionalArrays(lhs: lhsErrorLabels, rhs: rhsErrorLabels, cmp: { $0 < $1 })
+        case let (.bulkWriteError(writeErrors: lhsWriteErrors,
+                                  writeConcernError: lhsWCError,
+                                  result: _,
+                                  errorLabels: lhsErrorLabels),
+                  .bulkWriteError(writeErrors: rhsWriteErrors,
+                                  writeConcernError: rhsWCError,
+                                  result: _,
+                                  errorLabels: rhsErrorLabels)):
+            let cmp = { (l: BulkWriteError, r: BulkWriteError) in l.index < r.index }
+
+            return sortAndCompareOptionalArrays(lhs: lhsWriteErrors, rhs: rhsWriteErrors, cmp: cmp)
+                    && lhsWCError == rhsWCError
+                    && sortAndCompareOptionalArrays(lhs: lhsErrorLabels, rhs: rhsErrorLabels, cmp: { $0 < $1 })
+        default:
+            return false
+        }
+    }
+}
+
+extension UserError: Equatable {
+    public static func == (lhs: UserError, rhs: UserError) -> Bool {
+        switch (lhs, rhs) {
+        case (.logicError(message: _), .logicError(message: _)),
+             (.invalidArgumentError(message: _), .invalidArgumentError(message: _)):
+            return true
+        default:
+            return false
+        }
+    }
+}
+
+extension WriteError: Equatable {
+    public static func == (lhs: WriteError, rhs: WriteError) -> Bool {
+        return lhs.code == rhs.code
+    }
+}
+
+extension BulkWriteError: Equatable {
+    public static func == (lhs: BulkWriteError, rhs: BulkWriteError) -> Bool {
+        return lhs.code == rhs.code && lhs.index == rhs.index
```

Comment:
Is it worth checking the index here? Technically, this might consider two BulkWriteErrors equivalent because they have the same code and occurred on the same index within different `bulkWrite()` calls -- nothing about index ties the error back to a specific `bulkWrite()` operation, so that would make the equivalence check a bit coincidental.

Reply:
I agree, but this equality operator will only ever be used within our tests and as part of comparing two `.bulkWriteErrors` derived from the same request (note this is defined in `MongoError+Equatable.swift`, which is in our Testing folder). If this were part of the public API then I agree that we should omit the index, (or only conform to `Equatable` at the enum level), but for testing purposes, it is convenient to leave it in.

Reply:
Ah, I missed that this was in the test suite. SGTM!